packages/plugins/inline-trace/src/index.ts

```diff
@@ -0,0 +1,313 @@
+      if (!ctx) return
+
+      // result was already shipped (see InlineTraceContext.stopped)
+      if (ctx.stopped) {
+        return () => {
+          // noop
+        }
+      }
+
+      const node = newTraceNode(ctx, info.path)
+      node.type = info.returnType.toString()
+      node.parentType = info.parentType.toString()
+      node.startTime = hrTimeToDurationInNanos(process.hrtime(ctx.startHrTime))
+      if (
+        typeof info.path.key === 'string' &&
+        info.path.key !== info.fieldName
+      ) {
+        // field was aliased, send the original field name too
+        node.originalFieldName = info.fieldName
+      }
+
+      return () => {
+        node.endTime = hrTimeToDurationInNanos(process.hrtime(ctx.startHrTime))
+      }
+    },
+    onParse() {
+      return ({ context: { request }, result }) => {
+        const ctx = ctxForReq.get(request)
+        if (!ctx) return
+
+        if (result instanceof GraphQLError) {
+          handleErrors(ctx, [result], options.rewriteError)
+        } else if (result instanceof Error) {
+          handleErrors(
+            ctx,
+            [
+              new GraphQLError(result.message, {
+                originalError: result,
+              }),
+            ],
+            options.rewriteError,
+          )
+        }
+      }
+    },
+    onValidate() {
+      return ({ context: { request }, result: errors }) => {
+        if (errors.length) {
+          const ctx = ctxForReq.get(request)
+          if (ctx) handleErrors(ctx, errors, options.rewriteError)
+        }
+      }
+    },
+    onExecute() {
+      return {
+        onExecuteDone({
+          args: {
+            contextValue: { request },
+          },
+          result,
+        }) {
+          // TODO: should handle streaming results? how?
+          if (!isAsyncIterable(result) && result.errors?.length) {
+            const ctx = ctxForReq.get(request)
+            if (ctx) handleErrors(ctx, result.errors, options.rewriteError)
+          }
+        },
+      }
+    },
+    onResultProcess({ request, result }) {
+      const ctx = ctxForReq.get(request)
+      if (!ctx) return
+
+      // TODO: should handle streaming results? how?
+      if (isAsyncIterable(result)) return
+
+      if (result.extensions?.ftv1 !== undefined) {
+        throw new Error('The `ftv1` extension is already present')
+      }
+
+      // onResultProcess will be called only once since we disallow async iterables
+      if (ctx.stopped) throw new Error('Trace stopped multiple times')
+
+      ctx.stopped = true
+      ctx.trace.durationNs = hrTimeToDurationInNanos(
+        process.hrtime(ctx.startHrTime),
+      )
+      ctx.trace.endTime = nowTimestamp()
+
+      const encodedUint8Array = Trace.encode(ctx.trace).finish()
+      const encodedBuffer = Buffer.from(
+        encodedUint8Array,
+        encodedUint8Array.byteOffset,
+        encodedUint8Array.byteLength,
+      )
+
+      result.extensions = {
```

Comment:
Do you also think we can use `onExecuteDone`'s `setResult` for this?

Reply:
I used `onResultProcess` because I want the timing to end as late as possible, tracing the whole plugin chain.

Reply:
That's what I assumed, but I think `onExecuteDone` and `onResultProcess` are almost same in terms of timing, right?

Reply:
You're right. The only thing that can slow it up is other `onResultProcess` plugins that came before.

I'll move this whole thing to envelop then. ;)

Reply:
Actually, `onExecuteDone` cannot be used because it does not account for parsing and validation errors. See https://github.com/n1ru4l/envelop/issues/1491.